cmd/ci-operator/main_test.go

```diff
@@ -61,3 +76,204 @@ func TestSanitizeMessage(t *testing.T) {
+			Namespace: "some-namespace",
+		},
+		{
+			TestName:  "generate metadata with a custom metadata file",
+			Org:       "some-org",
+			Repo:      "some-repo",
+			ExtraOrg:  "some-extra-org",
+			ExtraRepo: "some-extra-repo",
+			ProwJobID: "some-prow-job-id",
+			Namespace: "some-namespace",
+			CustomMetadata: map[string]string{
+				"custom-field1": "custom-value1",
+				"custom-field2": "custom-value2",
+			},
+		},
+	}
+	for _, tc := range tests {
+		err := verifyMetadata(tc.Org, tc.Repo, tc.ExtraOrg, tc.ExtraRepo, tc.ProwJobID, tc.Namespace, tc.CustomMetadata)
+		if err != nil {
+			t.Errorf("Test case '%s': error while running test: %v", tc.TestName, err)
+		}
+	}
+}
+
+func verifyMetadata(org string,
+	repo string,
+	extraOrg string,
+	extraRepo string,
+	prowJobID string,
+	namespace string,
+	customMetadata map[string]string) error {
+	tempDir, err := ioutil.TempDir("", "")
+	if err != nil {
+		return fmt.Errorf("Unable to create temporary directory: %v", err)
+	}
+
+	defer os.RemoveAll(tempDir)
+
+	metadataFile := filepath.Join(tempDir, "metadata.json")
+
+	// Verify without custom metadata
+	o := createOptions(tempDir, org, repo, extraOrg, extraRepo, prowJobID, namespace)
+	err = o.writeMetadataJSON()
+	if err != nil {
+		return fmt.Errorf("error while writing metadata JSON: %v", err)
+	}
+
+	err = verifyMetadataFields(metadataFile, org, repo, extraOrg, extraRepo, prowJobID, namespace, map[string]string{})
+
+	info, err := os.Stat(metadataFile)
+	if err != nil {
+		return fmt.Errorf("unable to stat metadata file: %v", err)
+	}
+
+	originalModTime := info.ModTime()
+
+	testArtifactDirectory := filepath.Join(tempDir, "test-artifact-directory")
+	if os.Mkdir(testArtifactDirectory, os.FileMode(0755)) != nil {
+		return fmt.Errorf("unable to create artifact directory under temporary directory")
+	}
+
+	if len(customMetadata) > 0 {
+		testJSON, err := json.MarshalIndent(customMetadata, "", "")
+		if err != nil {
+			return fmt.Errorf("error marshalling custom metadata: %v", err)
+		}
+		err = ioutil.WriteFile(filepath.Join(testArtifactDirectory, "custom-prow-metadata.json"), []byte(testJSON), os.FileMode(0644))
+		if err != nil {
+			return fmt.Errorf("unable to create custom metadata file: %v", err)
+		}
+	}
+
+	// Write a bunch of empty files that should be ignored
+	err = ioutil.WriteFile(filepath.Join(testArtifactDirectory, "a-ignore1.txt"), []byte(""), os.FileMode(0644))
+	err = ioutil.WriteFile(filepath.Join(testArtifactDirectory, "b-ignore1.txt"), []byte(""), os.FileMode(0644))
+	err = ioutil.WriteFile(filepath.Join(testArtifactDirectory, "d-ignore1.txt"), []byte(""), os.FileMode(0644))
+	err = ioutil.WriteFile(filepath.Join(testArtifactDirectory, "e-ignore1.txt"), []byte(""), os.FileMode(0644))
+	if err != nil {
+		return fmt.Errorf("one or more of the empty *ignore files failed to write: %v", err)
+	}
+
+	// Sleep for one second to guarantee we get a different modified time of the second write
+	time.Sleep(1 * time.Second)
```

Comment:
Most of our unit tests run on the order of milliseconds or microseconds so something like this can have an outsize impact.

At least on my machine, `stat` gives me modification times down to the nanosecond, so we should be able to reduce this wait to 10ms or so. 

```
$ stat /tmp/
  File: /tmp/
  Size: 760       	Blocks: 0          IO Block: 4096   directory
Device: 2dh/45d	Inode: 22654       Links: 31
Access: (1777/drwxrwxrwt)  Uid: (    0/    root)   Gid: (    0/    root)
Context: system_u:object_r:tmp_t:s0
Access: 2019-11-07 10:09:19.414764485 -0800
Modify: 2019-11-07 11:42:24.296412280 -0800
Change: 2019-11-07 11:42:24.296412280 -0800
 Birth: -
```

However, adding time dependencies to a test is often buggy as you have no control over the kernel deciding to sleep your thread for whatever amount of time and in rare cases on time boundaries `time.After()` is not guaranteed to work. If we can validate that the code did something by checking for a changed field and not use time, that would be preferrable.

Reply:
In practice without the sleep the mod time before == the mod time after regardless of the write, I'm guessing due to CPU scheduling. I'm not sure how else to generically check that the file has or hasn't been modified other than adding a "revision" field or something to the metadata JSON itself. This is doable, it just sounds like it's not preferable.

Reply:
Thinking on it, I'll add in the revision. Should be straightforward.